File: gfg-solutions/DynamicProgramming/MaxLengthChain.cpp
// https://practice.geeksforgeeks.org/problems/max-length-chain/1

/*
The structure to use is as follows
struct val{
	int first;
	int second;
};*/

/*You are required to complete this method*/
bool compare(pair<int, int> p1, pair<int, int> p2){ 
    return (p1.second < p2.second); 
} 
int maxChainLen(struct val p[],int n){
    vector<pair<int, int> > a(n);
    for(int i = 0; i<n; i++){
        a[i] = make_pair(p[i].first, p[i].second);
    }
    
    sort(a.begin(), a.end(), compare);
    
    vector<int> dp(n, 1);
    int last = a[0].second;
    for(int i = 1; i<n; i++){
        if(a[i].first > last){
            last = a[i].second;
            dp[i] = dp[i-1] + 1;
        } else{
            dp[i] = dp[i-1];
        }
    }
    return dp[n-1];
}

File: gfg-leetcode-solutions/DynamicProgramming/MaxLengthChain.java
// https://practice.geeksforgeeks.org/problems/max-length-chain/1

class CompareBySecond implements Comparator<Pair>
{
   public int compare(Pair a, Pair b)
    {
        return a.y - b.y;
    }
}

class GfG
{
    int maxChainLength(Pair arr[], int n)
    {
        Arrays.sort(arr, new CompareBySecond());
        
        int dp[] = new int[n];
        Arrays.fill(dp, 1);
        
        int last = arr[0].y;
        for(int i = 1; i<n; i++){
            if(arr[i].x > last){
                dp[i] = dp[i-1]+1;
                last = arr[i].y;
            }
            else
                dp[i] = dp[i-1];
        }
        return dp[n-1];
    }
}